pkg/server/test/check.go

```diff
@@ -388,3 +389,74 @@ func BenchmarkCheckWithTrace(b *testing.B, datastore storage.OpenFGADatastore) {
 
 	checkResponse = r
 }
+
+// BenchmarkCheckWithDirectResolution ensures that when the user can be found through direct resolution,
+// the code that tries to find the user through usersets is aborted immediately.
+func BenchmarkCheckWithDirectResolution(b *testing.B, datastore storage.OpenFGADatastore) {
+	ctx := context.Background()
+	tracer := telemetry.NewNoopTracer()
+	meter := telemetry.NewNoopMeter()
+	logger := logger.NewNoopLogger()
+	store := ulid.Make().String()
+
+	data, err := os.ReadFile(gitHubTestDataFile)
+	require.NoError(b, err)
+
+	var gitHubTypeDefinitions openfgapb.WriteAuthorizationModelRequest
+	err = protojson.Unmarshal(data, &gitHubTypeDefinitions)
+	require.NoError(b, err)
+
+	model := &openfgapb.AuthorizationModel{
+		Id:              ulid.Make().String(),
+		SchemaVersion:   typesystem.SchemaVersion1_0,
+		TypeDefinitions: gitHubTypeDefinitions.GetTypeDefinitions(),
+	}
+
+	err = datastore.WriteAuthorizationModel(ctx, store, model)
+	require.NoError(b, err)
+
+	// add user to many usersets
+	for i := 0; i < 1000; i++ {
+		err = datastore.Write(ctx, store, []*openfgapb.TupleKey{}, []*openfgapb.TupleKey{{
+			User:     "anne",
+			Relation: "member",
+			Object:   fmt.Sprintf("team:%d", i),
+		}})
+		require.NoError(b, err)
+	}
+
+	// one of those usersets gives access to the repo
+	err = datastore.Write(ctx, store, []*openfgapb.TupleKey{}, []*openfgapb.TupleKey{{
+		User:     "team:999#member",
+		Relation: "owner",
+		Object:   "repo:openfga",
+	}})
+	require.NoError(b, err)
+
+	// add direct access to the repo
+	err = datastore.Write(ctx, store, []*openfgapb.TupleKey{}, []*openfgapb.TupleKey{{
+		User:     "anne",
+		Relation: "owner",
+		Object:   "repo:openfga",
+	}})
+	require.NoError(b, err)
+
+	checkQuery := commands.NewCheckQuery(datastore, tracer, meter, logger, defaultResolveNodeLimit)
```

Comment:
Instead of using the `command` can you please use the server instead?

```
s := &Server{...}

resp, err := s.Check(...)
```

This way this test is portable to whatever implementation details implement `Check`. As it stands this won't translate if we refactor to use a different Check implementation, because it's only testing the `CheckQuery` command.

Reply:
I don't understand

> because it's only testing the CheckQuery command.

This is what we want, no?

`s.Check` delegates to `commands.NewCheckQuery`.

Reply:
Nah, we want to check `s.Check`. If we only test the command then if `s.Check` implementation changes we won't necessarily be testing what we hoped to be testing. Instead of testing the command, test the RPC endpoint.

Reply:
I tried to make that change and I couldn't - I get a circular reference error.

But in any case

> if s.Check implementation changes

https://github.com/openfga/openfga/blob/33a5ef166936e916f540d0381048b1143363cf0b/pkg/server/server.go#L215

`s.Check` is supposed to be, and is, a delegator to the command. This shouldn't change, IMO.

Reply:
The other two benchmarks also use `CheckCommand`, so if we are going to make this change, we should send a different PR where we update them all, IMO

Reply:
☝️ I agree with Maria here. If we want let's refactor all of them in a future PR.